src/index.ts

```diff
@@ -148,30 +190,77 @@ export class CSSStyleObserver {
   }
 
   /**
-   * Collect CSS variable values and invoke callback.
+   * Process observed variables and collect changes
+   *
+   * @param computedStyle The computed style of the target element
+   * @param processChange Function to process each changed property
+   * @param finalizeChanges Function to finalize and handle collected changes
    */
-  private _handleUpdate(): void {
-    if (this._targetElement) {
-      const computedStyle = getComputedStyle(this._targetElement);
-
-      const changedProperties: CSSDeclarations = {};
-
-      this._observedVariables.forEach(propertyName => {
+  private _processChanges(
+    computedStyle: CSSStyleDeclaration,
+    processChange: (propertyName: string, currentValue: string, previousValue: string, hasChanged: boolean) => void,
+    finalizeChanges: () => void
+  ) {
+      this._observedVariables.forEach((propertyName) => {
         const currentValue = computedStyle.getPropertyValue(propertyName);
         const previousValue = this._cachedValues[propertyName] || '';
+        const hasChanged = currentValue !== previousValue;
+
+      if (this._notificationMode === NotificationMode.ALL || hasChanged) {
+        processChange(propertyName, currentValue, previousValue, hasChanged);
+        this._cachedValues[propertyName] = currentValue;
+      }
+    });
+
+    finalizeChanges();
```

Comment:
I think it’d be better to simply return the results here.

Also see the “This is duplicated in both handlers that are currently supported.” comment further down. By returning here, you can dedupe some code.

Reply:
The changes are now returned directly from the process and I’ve removed the `finalizeChanges` function to simplify the code